Added accelerator edit page

Gkrumbach07 opened the pull request:
<!--- If this is a non-code change, this template is not required; reference any issues or top-level descriptions as needed -->
<!--- All code change PRs should relate to an issue, reference it here; see example below -->
<!--- Closes: #123 -->
closes: https://github.com/opendatahub-io/odh-dashboard/issues/1375

## Description
<!--- Describe your changes in detail; the what, the why, any findings, etc -->
<!--- Include any screenshots of changed UI; Include any gifs if it was a flow / UX change -->

- created a accelerator routing component to cover all the pages
- broke manage accelerator into a edit component and the core component
- edit fetches the existing profile and handles any errors there
- core manage page is a basic form with
   - name, identifier, description, enabled, and tolerations
   - tolerations is a table
       - adding new/editing tolerations opens a modal with toleration fields. some are drop downs, some are text, key is required, and seconds is a radio that can be set to forever or a number
       - toleration modal does not make network calls
       - toleration modal has non blocking info alerts when invalid combo of inputs is detected
- new k8s api functions for update, get, and post
- new function for generating k8s yaml

**Error**
<img width="1290" alt="Screenshot 2023-11-01 at 2 49 19 PM" src="https://github.com/opendatahub-io/odh-dashboard/assets/12587674/436575f8-6c9d-4b5c-8ae3-70f3af19e732">

**Accelerator Page**
![Screenshot 2023-11-03 at 1 52 39 PM](https://github.com/opendatahub-io/odh-dashboard/assets/12587674/b95f2d41-b306-4a3d-a9df-7b396696d41c)

**Helper texts on accelerator page**
![Screenshot 2023-11-03 at 1 52 49 PM](https://github.com/opendatahub-io/odh-dashboard/assets/12587674/28b44dcb-ca45-40c4-9c04-592b9ca512b6)
<img width="647" alt="image" src="https://github.com/opendatahub-io/odh-dashboard/assets/12587674/7450220a-adfd-4a3e-ba51-c502d63294e4">


**Toleration Modal**
![Screenshot 2023-11-03 at 1 53 03 PM](https://github.com/opendatahub-io/odh-dashboard/assets/12587674/32381855-ffaa-480d-ba5f-93c93f7f11d8)
![Screenshot 2023-11-03 at 1 53 11 PM](https://github.com/opendatahub-io/odh-dashboard/assets/12587674/cc808087-5caf-49ac-8486-9298925ec5ba)
![Screenshot 2023-11-03 at 1 53 18 PM](https://github.com/opendatahub-io/odh-dashboard/assets/12587674/4d8302d0-3761-4cfd-bf3b-91b2a77607bb)
![Screenshot 2023-11-03 at 1 53 30 PM](https://github.com/opendatahub-io/odh-dashboard/assets/12587674/f2582410-c2c1-43c9-83f5-7a53fa4889bd)
![Screenshot 2023-11-03 at 1 54 46 PM](https://github.com/opendatahub-io/odh-dashboard/assets/12587674/8c9beb11-3660-467c-aeaf-c310d3b3ecc0)


## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

- Go to /acceleratorProfiles/manage to open the create page
- play around with the form
   - add and remove tolerations from the table
   - make sure toleration form is reflected in the core form
   - submit when satisfied with form
- edit the accelerator by going to /acceleratorProfiles/manage/{resource-id}
   - edit the name to something else
   - remove tolerations
   - edit other fields
   - submit and edit it again. make sure all is reflected
- edit a profile that does not exist, see error page
- go to routes such as /acceleratorProfiles/wrong or /acceleratorProfiles/manage/id/not/a/route and make sure you are redirected to /acceleratorProfiles

## Test Impact
<!--- What tests have you done to covert implemented functionality -->
<!--- If tests are not applicable, explain why here -->
tests exist for the page in all three states (create, edit, error)

## Request review criteria:
<!--- This PR will be merged by any repository approver when it meets all the points in the checklist -->
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->

Self checklist (all need to be checked):
- [x] The developer has manually tested the changes and verified that the changes work
- [x] Commits have been squashed into descriptive, self-contained units of work (e.g. 'WIP' and 'Implements feedback' style messages have been removed)
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has added tests or explained why testing cannot be added (unit tests & storybook for related changes)

If you have UI changes: 
<!--- You can ignore these if you are doing manifest, backend, internal logic, etc changes; aka non-UI / visual changes -->
- [x] Included any necessary screenshots or gifs if it was a UI change.
- [x] Included tags to the UX team if it was a UI/UX change (find relevant UX in the [SMEs](https://github.com/opendatahub-io/odh-dashboard/tree/main/docs/smes.md) section).

After the PR is posted & before it merges:
- [ ] The developer has tested their solution on a cluster by using the image produced by the PR to `main`

Gkrumbach07 commented:
## UX Context
@kaedward can you review the wording on this PR. specifically there is the
-  form header
- toleration table empty state
- identifier helper text (or any other helper text you think should exist)
- toleration form helper texts (key has one, but others could too?)
- all the dropdown item descriptions for toleration form
- the two info alerts in the toleration form
- the helper (?) over the table heading columns

for tolerations i used this [k8s api docs](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.24/#toleration-v1-core) schema for tolerations. they mention some things that i translated over

Gkrumbach07 commented:
@yannnz and @xianli123 Here is the accelerator edit/create page. some things
- is the route name ok? `/acceleratorProfiles/manage/{cr-name}`
- thats it.

kaedward commented:
@Gkrumbach07 I left comments addressing all of the items in your list in the mocks (plus a few other items that you didn't list, I believe), except for the toleration form helper texts. I was having a hard time finding reference to the key helper text in the documentation you linked. I'm not sure what that one is conveying, so I can't tell what we'd want to use for the other fields. Please LMK if any of my suggestions are inaccurate - it might be easier to walk through these together.

dpanshug commented on frontend/src/pages/acceleratorProfiles/screens/manage/tolerations/TolerationRow.tsx:
```diff
@@ -0,0 +1,40 @@
+      <Td dataLabel="Effect">{formatValue(toleration.effect)}</Td>
+      <Td dataLabel="Toleration Seconds">{formatSeconds(toleration.tolerationSeconds)}</Td>
+      <Td isActionCell>
+        <ActionsColumn
```
```suggestion
        <ActionsColumn
         menuAppendTo="parent"
```

Gkrumbach07 replied:
that is standard behavior for any dropdown, action menu, etc... the reason its weird here is that it is at the bottom. the solution is for the user to scroll a bit to see the delete. it is a bit weird here because you cant scroll enough to see the full menu... but setting that to parent causes the element to go over the footer like a modal would. it would be better to just have the menu go up instead of down
![Screenshot 2023-11-03 at 3 19 53 PM](https://github.com/opendatahub-io/odh-dashboard/assets/12587674/a1e7782b-0a7f-4bec-ac62-a914d8788499)

Gkrumbach07 replied:
I think it looks off. @yannnz what would you rather see. the menu covering the footer, or the menu opening above the button

yannnz replied:
I vote for the menu opening above the button. The same reason as you suggested @Gkrumbach07. cc: @dpanshug

dpanshug commented on frontend/src/pages/acceleratorProfiles/screens/manage/tolerations/ManageTolerationModal.tsx:
```diff
@@ -0,0 +1,84 @@
+  };
+
+  return (
+    <Modal
+      title="Manage Toleration"
```
When editing the toleration, should we make the modal title something like `Edit toleration`?

Gkrumbach07 commented:
@kaedward @xianli123 @yannnz

I updated the screenshots, let me know if there is anything that needs fixing

yannnz commented:
I am a bit confused of the way we use Tooltips vs Popovers in RHODS/ODH. Most of the screenshots use Popovers with the question icons, except the one for Toleration data list column, which uses Tooltip. Is there a reason for this? 
![image](https://github.com/opendatahub-io/odh-dashboard/assets/10469467/7d6b14e3-8202-4eed-9230-4cbe318bb1a0)
Here is the PF guide of when to use tooltips vs popovers. At the bottom, it is suggested not to use tooltips with question-circle icons to present contextual information in forms and other areas. Is it different from the convention of RHODS /ODH?

Gkrumbach07 commented:
@yannnz no, it should be popover. i have updated it. But should there even be helpers here? Because there are already helping text in the toleration modal form

Gkrumbach07 commented:
@dpanshug Ive updated with changes. can you take another look

yannnz commented:
> @yannnz no, it should be popover. i have updated it. But should there even be helpers here? Because there are already helping text in the toleration modal form

It's a good questions. Seems duplicated with the modal form.

dpanshug commented on frontend/src/pages/acceleratorProfiles/AcceleratorProfilesRoutes.tsx:
```diff
@@ -0,0 +1,16 @@
+import AcceleratorProfiles from './screens/list/AcceleratorProfiles';
+import ManageAcceleratorProfile from './screens/manage/ManageAcceleratorProfile';
+
+const AcceleratorProfilesRoutes: React.FC = () => (
+  <Routes>
+    <Route path="/" element={<AcceleratorProfiles />} />
+    <Route path="/manage" element={<ManageAcceleratorProfile />} />
+    <Route path="/manage/:acceleratorProfileName" element={<EditManageAcceleratorProfile />} />
+    <Route path="*" element={<Navigate to="." />} />
+  </Routes>
+);
```
When navigating to a URL like `acceleratorProfiles?abcd`, it opens the page for `/acceleratorProfiles` but does not update the URL to reflect the change. However, when using a format like `acceleratorProfiles/abcd`, the URL is appropriately updated to `/acceleratorProfiles`. Is this behavior expected or normal?

![image](https://github.com/opendatahub-io/odh-dashboard/assets/97534722/00e0d51c-6419-465b-ab16-a7926de26df8)

Gkrumbach07 replied:
Yes, the behavior you're observing is expected and normal due to the way URLs and routing are handled on the web.

When you navigate to a URL that includes `?abcd`, the part `?abcd` is interpreted as a query string. Query strings do not change the route; they are just additional data for the server to process while loading the same route. Hence, the URL does not change from `/acceleratorProfiles` because the query string is not considered a new route, just an addition to the current route.

On the other hand, when you use the format `acceleratorProfiles/abcd`, the part `/abcd` is interpreted as a new path in the URL, which is a different route from just `/acceleratorProfiles`. If our router encounters a route it doesn't recognize, it redirects to a fallback route, which in our case is `/acceleratorProfiles`. So when you navigate to `acceleratorProfiles/abcd`, the URL is replaced with the new route because `/abcd` is considered part of the path and not just additional data like a query string.

dpanshug replied:
Great, thanks for the explanation.

Gkrumbach07 commented on frontend/src/pages/acceleratorProfiles/screens/manage/ManageAcceleratorProfileDetailsSection.tsx:
```diff
@@ -0,0 +1,101 @@
+  >
+    <Stack hasGutter>
+      <StackItem>
+        <FormGroup
+          label="Name"
+          labelIcon={
+            <Popover bodyContent="An identifier is a unique string that names a specific hardware accelerator resource.">
```
good catch

Gkrumbach07 replied:
fixed

Gkrumbach07 commented on backend/src/routes/api/accelerator-profiles/acceleratorProfilesUtils.ts:
```diff
@@ -0,0 +1,170 @@
+    }
+
+    // Update the modified date annotation
+    currentProfile.metadata.annotations['opendatahub.io/modified-date'] = new Date().toISOString();
```
this happens when annotations is undefined. i added a check to add that to the object before setting the date

Gkrumbach07 replied:
this should be fixed now

dpanshug commented:
/lgtm

Gkrumbach07 commented:
Did a final review myself and found a few minor things
- removed unused functions
- renamed a component to be more clear
- switched the route for `/add` to `/create` because the UX uses the wording create and not add
- cleaned up duplicate code
- added removeFindDomNode to popovers
- minor text changes in errors
- added roles for service account to create, delete, and patch

all tests pass and code has been tested manually.

DaoDaoNoCode reviewed (commented):
Overall it works well, I left some comments about code quality and a small UX issue.

DaoDaoNoCode reviewed (approved):
/lgtm
Thanks for the changes

Gkrumbach07 commented:
/approve
